File: app/services/email_service.py
from flask import render_template
from flask_mail import Message
from app.extensions import mail
from app.extensions import db
from app.models.user import User
from app.models.user_setting import UserSetting


def send_html_email(recipient_email, subject, template_name, **context):
    from flask import current_app as app
    if not app.config['MAIL_USERNAME'] or not app.config['MAIL_PASSWORD']:
        return
    try:
        html = render_template(f'email/{template_name}', **context)
        msg = Message(subject, recipients=[recipient_email], html=html)
        mail.send(msg)
    except Exception as e:
        print(f"Email send failed to {recipient_email}: {e}")


def send_registration_confirmation(student, event):
    organizer = User.query.get(event.created_by)
    organizer_name = organizer.name if organizer else 'Organizer'
    send_html_email(
        student.email,
        f"✅ You're registered for {event.title}",
        'event_registration_confirmation.html',
        student=student,
        event=event,
        organizer_name=organizer_name
    )


def send_event_update_notification(user, event, change_description):
    settings = UserSetting.query.filter_by(user_id=user.id).first()
    if settings and not settings.email_notifications:
        return
    send_html_email(
        user.email,
        f"📋 Update: {event.title}",
        'event_update_notification.html',
        user=user,
        event=event,
        change_description=change_description
    )


def send_password_reset_notification(user, forced=False):
    settings = UserSetting.query.filter_by(user_id=user.id).first()
    if settings and not settings.email_notifications:
        return
    subject = '🔐 Password Reset - CampusCore' if not forced else '🔐 Password Reset Required - CampusCore'
    send_html_email(
        user.email,
        subject,
        'password_reset_notification.html',
        user=user,
        forced=forced
    )


def send_account_update_notification(user, changes):
    settings = UserSetting.query.filter_by(user_id=user.id).first()
    if settings and not settings.email_notifications:
        return
    send_html_email(
        user.email,
        f"🔔 Your CampusCore Account Has Been Updated",
        'account_update_notification.html',
        user=user,
        changes=changes
    )


def send_new_user_notification(new_user):
    admins = User.query.filter_by(role='admin').all()
    for admin in admins:
        settings = UserSetting.query.filter_by(user_id=admin.id).first()
        if settings and not settings.email_notifications:
            continue
        send_html_email(
            admin.email,
            f"👤 New User Registration: {new_user.name}",
            'new_user_notification.html',
            admin=admin,
            new_user=new_user
        )

File: services/emailService.js
const transporter = require('../config/mailer');
const templates = require('./emailTemplates');
const EmailLog = require('../models/EmailLog');

const send = async ({ to, subject, html, type }) => {
  const recipients = Array.isArray(to) ? to.join(', ') : to;

  const mailOptions = {
    from: `"${process.env.EMAIL_FROM_NAME}" <${process.env.EMAIL_FROM}>`,
    to: recipients,
    subject,
    html,
  };

  try {
    await transporter.sendMail(mailOptions);

    await EmailLog.create({
      recipientEmail: recipients,
      subject,
      type,
      status: 'sent',
      sentAt: new Date(),
    });

    console.log(`[EMAIL] ✅ ${type} → ${recipients}`);
  } catch (error) {
    await EmailLog.create({
      recipientEmail: recipients,
      subject,
      type,
      status: 'failed',
      error: error.message,
      sentAt: new Date(),
    });

    console.error(`[EMAIL] ❌ ${type} failed → ${error.message}`);
  }
};

const sendIfEnabled = async (user, sendFn, skipPreferenceCheck = false) => {
  if (!skipPreferenceCheck && user.preferences?.emailNotifications === false) {
    console.log(`[EMAIL] Skipped — ${user.email} opted out.`);
    return;
  }
  return sendFn();
};

exports.sendEventRegistrationConfirmation = async (student, event) => {
  await sendIfEnabled(student, () =>
    send({
      to: student.email,
      subject: `✅ You're registered for ${event.title} — CampusCore`,
      html: templates.eventRegistrationConfirmation({ student, event }),
      type: 'EVENT_REGISTRATION_CONFIRMATION',
    })
  );
};

exports.sendEventUpdateNotification = async (students, event, changes) => {
  const eligible = students.filter(s => s.preferences?.emailNotifications !== false);

  for (const student of eligible) {
    await send({
      to: student.email,
      subject: `🔔 Update for "${event.title}" — CampusCore`,
      html: templates.eventUpdateNotification({ event, changes }),
      type: 'EVENT_UPDATE_NOTIFICATION',
    });
  }
};

exports.sendOrganizerEditAuditToAdmins = async (admins, actingAdmin, organizer, changes) => {
  for (const admin of admins) {
    await send({
      to: admin.email,
      subject: `🛡️ Audit: ${actingAdmin.firstName} edited ${organizer.firstName}'s profile`,
      html: templates.organizerEditAuditForAdmins({ actingAdmin, organizer, changes }),
      type: 'ORGANIZER_EDIT_AUDIT_ADMIN',
    });
  }
};

exports.sendProfileEditedNotificationToOrganizer = async (organizer, actingAdmin, changes) => {
  await sendIfEnabled(organizer, () =>
    send({
      to: organizer.email,
      subject: `📝 Your CampusCore profile was updated by an Admin`,
      html: templates.profileEditedForOrganizer({ organizer, actingAdmin, changes }),
      type: 'PROFILE_EDITED_ORGANIZER',
    })
  );
};

exports.sendForcePasswordResetNotification = async (user, actingAdmin, resetLink) => {
  await sendIfEnabled(
    user,
    () => send({
      to: user.email,
      subject: `🔐 Your CampusCore password was reset by an Administrator`,
      html: templates.forcePasswordReset({ user, actingAdmin, resetLink }),
      type: 'FORCE_PASSWORD_RESET',
    }),
    true
  );
};

exports.sendNewUserRegistrationAlertToAdmins = async (admins, newUser) => {
  for (const admin of admins) {
    await send({
      to: admin.email,
      subject: `👤 New ${newUser.role} registered — ${newUser.firstName} ${newUser.lastName}`,
      html: templates.newUserRegistrationAlert({ newUser }),
      type: 'NEW_USER_REGISTRATION_ALERT',
    });
  }
};
